src/Common/src/Interop/Windows/winhttp/Interop.SafeWinHttpHandle.cs

```diff
@@ -30,6 +30,31 @@ public static void DisposeAndClearHandle(ref SafeWinHttpHandle winHttpHandler)
                 }
             }
 
+            // The SafeHandle base class is inconsistent w.r.t. how Dispose() and DangerousRelease()
+            // are implemented. Both methods can cause the ref count on the SafeHandle to be decremented.
+            // And when the ref count reaches zero, ReleaseHandle() is called and the handle is marked
+            // as 'closed'. However, there is a separate internal state bit to track Dispose() calls.
+            // Calling Dispose() more than once is a no-op.  However, calling DangerousRelease() and then
+            // Dispose() will cause ObjectDisposedException to be thrown if the previous DangerousRelease()
+            // caused the handle to be closed. This seems inconsistent especially since both methods can
+            // decrement the ref count and cause ReleaseHandle() to be called.
```

Comment:
> Right now, there is a single SafeWinHttpHandle class. But it sounds like there needs to a variant of this class that can take a "parent" SafeHandle in order to manage AddRef/Release on it as you described. 

Ah, ok, I guess that's the part I was missing, that we were working within the constraints of the SafeHandle types already employed.  Yeah, give the approach I suggested a try, and let me know how it goes.